File: receivers/native/generic-core-3/src/test/c++/zscript/semanticParser/SemanticParserMulticommandTest.cpp
/*
 * Zscript Library - Command System for Microcontrollers)
 * Copyright (c) 2022 Zscript team (Susan Witts, Alicia Witts)
 *
 * SPDX-License-Identifier: MIT
 */

#include <iostream>
#include "../../../../main/c++/zscript/modules/core/CoreModule.hpp"
#include "../../../../main/c++/zscript/semanticParser/SemanticParser.hpp"
#include "../../../../main/c++/zscript/notifications/ZscriptNotificationSource.hpp"
#include "../../../../main/c++/zscript/tokenizer/ZscriptTokenizer.hpp"
#include "../../../../main/c++/zscript/Zscript.hpp"
#include "BufferOutStream.hpp"
#include "ParserActionTester.hpp"
namespace Zscript {
namespace GenericCore {
class zp {
    static uint16_t currentRnd;

public:
    static const uint8_t lockByteCount = 3;

    static uint16_t generateRandom16() {
        currentRnd++;
        currentRnd ^= currentRnd * 23;
        return currentRnd;
    }
};
uint16_t zp::currentRnd = 1249;

class SemanticParserTest {
    static const SemanticActionType INVALID = SemanticActionType::INVALID;
    static const SemanticActionType RUN_COMMAND = SemanticActionType::RUN_COMMAND;
    static const SemanticActionType RUN_FIRST_COMMAND = SemanticActionType::RUN_FIRST_COMMAND;
    static const SemanticActionType WAIT_FOR_TOKENS = SemanticActionType::WAIT_FOR_TOKENS;
    static const SemanticActionType END_SEQUENCE = SemanticActionType::END_SEQUENCE;
    static const SemanticActionType CLOSE_PAREN = SemanticActionType::CLOSE_PAREN;
    static const SemanticActionType ERROR = SemanticActionType::ERROR;
    uint8_t data[256];
    TokenRingBuffer<zp> buffer;

    SemanticParser<zp> parser;

    ZscriptTokenizer<zp> tokenizer;

    BufferOutStream<zp> outStream;

    void feedToTokenizer(const char *text) {
        for (uint16_t i = 0; text[i] != 0; ++i) {
            tokenizer.accept(text[i]);
        }
    }
    void checkAgainstOut(const char *text) {
        DataArrayWLeng16 data = outStream.getData();
        uint16_t textLen = 0;
        for (uint16_t i = 0; text[i] != 0; ++i) {
            textLen++;
        }
        bool worked = true;
        if (textLen != data.length) {
            std::cerr << "Out length doesn't match expected (Exp Len: " << textLen << ", actual: " << data.length << ")\n";
            worked = false;
        } else {
            for (uint16_t i = 0; text[i] != 0; ++i) {
                if (data.data[i] != text[i]) {
                    std::cerr << "Out value doesn't match expected\n ";
                    worked = false;
                }
            }
        }
        if (!worked) {
            std::cerr << "Expected: " << text;
            std::cerr << "\nActual: ";
            for (uint16_t i = 0; i < data.length; ++i) {
                std::cerr << data.data[i];
            }
            std::cerr << "\n";
            throw 0;
        }
    }

    void checkParserState(State s) {
        if (parser.getState() != s) {
            std::cerr << "Bad parser state\n";
            throw 0;
        }
    }

    void checkActionType(ZscriptAction<zp> a, SemanticActionType t) {
        if ((SemanticActionType) a.getType() != t) {
            std::cerr << "Bad action type: \nExpected: " << (uint8_t) t << "\nActual: " << (uint8_t) a.getType() << "\n";
            throw 0;
        }
    }

    void shouldHandleActionTypesAndIO(const char *input, const char *output) {
        ParserActionTester<zp> parserActionTester(&Zscript<zp>::zscript, &buffer, &tokenizer, &parser, &outStream);

        parserActionTester.parseSnippet(input, output);
        checkAgainstOut(output);
        if (outStream.isOpen()) {
            std::cerr << "Out stream open unexpectedly\n";
            throw 0;
        }
    }
public:
    SemanticParserTest() :
            buffer(data, 256), parser(&buffer), tokenizer(&buffer, 2) {
    }
    void shouldHandleTwoEmptyCommands() {
        ParserActionTester<zp> parserActionTester(&Zscript<zp>::zscript, &buffer, &tokenizer, &parser, &outStream);
        SemanticActionType types[] = { RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, WAIT_FOR_TOKENS };

        parserActionTester.parseSnippet("\n \n", types, 5);
        checkAgainstOut("!\n!\n");
        if (outStream.isOpen()) {
            std::cerr << "Out stream open unexpectedly\n";
            throw 0;
        }
    }

    void shouldHandleSeveralEmptyCommands() {
        ParserActionTester<zp> parserActionTester(&Zscript<zp>::zscript, &buffer, &tokenizer, &parser, &outStream);
        SemanticActionType types[] = { RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE,
                WAIT_FOR_TOKENS };

        parserActionTester.parseSnippet("\n\n \n\n", types, 9);
        checkAgainstOut("!\n!\n!\n!\n");
        if (outStream.isOpen()) {
            std::cerr << "Out stream open unexpectedly\n";
            throw 0;
        }
    }

    void shouldHandleSeveralBasicCommands() {
        ParserActionTester<zp> parserActionTester(&Zscript<zp>::zscript, &buffer, &tokenizer, &parser, &outStream);
        SemanticActionType types[] = { RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE,
                WAIT_FOR_TOKENS };

        parserActionTester.parseSnippet("Z1A\nZ1B\n Z1C\nZ1D\n", types, 9);
        checkAgainstOut("!AS\n!BS\n!CS\n!DS\n");
        if (outStream.isOpen()) {
            std::cerr << "Out stream open unexpectedly\n";
            throw 0;
        }
    }

    static void shouldProduceMultipleResponses() {
        SemanticParserTest s1;
        s1.shouldHandleActionTypesAndIO("Z1 A2 +1234\n Z1 A3B3 +5678\n", "!A2+1234S\n!A3B3+5678S\n");
        SemanticParserTest s2;
        s2.shouldHandleActionTypesAndIO("Z1A S1 \n Z1B\n", "!AS1\n!BS\n");
        SemanticParserTest s3;
        s3.shouldHandleActionTypesAndIO("Z1A S1 | Z1B S1 \n Z1C\n", "!AS1|BS1\n!CS\n");
        SemanticParserTest s4;
        s4.shouldHandleActionTypesAndIO("Z1A S13 \n Z1B\n", "!AS13\n!BS\n");
        SemanticParserTest s5;
        s5.shouldHandleActionTypesAndIO("Z1A S13 | Z1B S1 \n Z1C\n", "!AS13\n!CS\n");
    }

    static void shouldProduceMultipleResponsesWithComments() {
        SemanticParserTest s1;
        s1.shouldHandleActionTypesAndIO("Z1 A \n#hi\n", "!AS\n");
        SemanticParserTest s2;
        s2.shouldHandleActionTypesAndIO("Z1 A \n#hi\n#abc\nZ1B\n", "!AS\n!BS\n");
    }

    static void shouldProduceMultipleResponsesWithAddressing() {
        // Addressing actions are passed on downstream, not acknowledged - they should not appear in the response.
        SemanticParserTest s1;
        s1.shouldHandleActionTypesAndIO("Z1 A \n@1 Z1B\n", "!AS\n");
        SemanticParserTest s2;
        s2.shouldHandleActionTypesAndIO("Z1 A \n@1.2 Z1B\n@2.2 Z1C\n Z1D\n", "!AS\n!DS\n");
    }

};

}
}

int main(int argc, char **argv) {
    (void) argc;
    (void) argv;

    Zscript::GenericCore::SemanticParserTest s1;
    s1.shouldHandleTwoEmptyCommands();

    Zscript::GenericCore::SemanticParserTest s2;
    s2.shouldHandleSeveralEmptyCommands();

    Zscript::GenericCore::SemanticParserTest s3;
    s3.shouldHandleSeveralBasicCommands();

    Zscript::GenericCore::SemanticParserTest::shouldProduceMultipleResponses();

    Zscript::GenericCore::SemanticParserTest::shouldProduceMultipleResponsesWithComments();

    Zscript::GenericCore::SemanticParserTest::shouldProduceMultipleResponsesWithAddressing();
}

File: receivers/jvm/java-receiver/src/test/java/net/zscript/javareceiver/semanticParser/SemanticParserMulticommandTest.java
package net.zscript.javareceiver.semanticParser;

import java.io.IOException;
import java.util.List;
import java.util.stream.Stream;

import static net.zscript.javareceiver.semanticParser.SemanticAction.ActionType.END_SEQUENCE;
import static net.zscript.javareceiver.semanticParser.SemanticAction.ActionType.RUN_FIRST_COMMAND;
import static net.zscript.javareceiver.semanticParser.SemanticAction.ActionType.WAIT_FOR_TOKENS;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.params.provider.Arguments.of;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import net.zscript.javareceiver.core.StringWriterOutStream;
import net.zscript.javareceiver.core.Zscript;
import net.zscript.javareceiver.modules.core.ZscriptCoreModule;
import net.zscript.tokenizer.TokenBuffer;
import net.zscript.tokenizer.TokenRingBuffer;
import net.zscript.tokenizer.Tokenizer;

class SemanticParserMulticommandTest {
    private final TokenBuffer buffer    = TokenRingBuffer.createBufferWithCapacity(256);
    private final Tokenizer   tokenizer = new Tokenizer(buffer.getTokenWriter(), 2);

    private final SemanticParser parser = new SemanticParser(buffer.getTokenReader(), new ExecutionActionFactory());

    private final Zscript zscript = new Zscript();

    private StringWriterOutStream outStream;
    private ParserActionTester    parserActionTester;

    @BeforeEach
    void setup() throws IOException {
        zscript.addModule(new ZscriptCoreModule());
        outStream = new StringWriterOutStream();
        parserActionTester = new ParserActionTester(zscript, buffer, tokenizer, parser, outStream);
    }

    @Test
    public void shouldHandleTwoEmptyCommands() {
        parserActionTester.parseSnippet("\n \n", List.of(RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, WAIT_FOR_TOKENS));
        assertThat(outStream.getString()).isEqualTo("!\n!\n");
        assertThat(outStream.isOpen()).isFalse();
    }

    @Test
    public void shouldHandleSeveralEmptyCommands() {
        parserActionTester.parseSnippet("\n\n \n\n",
                List.of(RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, WAIT_FOR_TOKENS));
        assertThat(outStream.getString()).isEqualTo("!\n!\n!\n!\n");
        assertThat(outStream.isOpen()).isFalse();
    }

    @Test
    public void shouldHandleSeveralBasicCommands() {
        parserActionTester.parseSnippet("Z1A\nZ1B\n Z1C\nZ1D\n",
                List.of(RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, RUN_FIRST_COMMAND, END_SEQUENCE, WAIT_FOR_TOKENS));
        assertThat(outStream.getString()).isEqualTo("!AS\n!BS\n!CS\n!DS\n");
        assertThat(outStream.isOpen()).isFalse();
    }

    @ParameterizedTest
    @MethodSource
    public void shouldProduceMultipleResponses(final String text, String finalOutput) {
        parserActionTester.parseSnippet(text, finalOutput);
        assertThat(outStream.isOpen()).isFalse();
    }

    private static Stream<Arguments> shouldProduceMultipleResponses() {
        return Stream.of(
                of("Z1 A2 +1234\n Z1 A3B3 +5678\n", "!A2+1234S\n!A3B3+5678S\n"),
                of("Z1A S1 \n Z1B\n", "!AS1\n!BS\n"),
                of("Z1A S1 | Z1B S1 \n Z1C\n", "!AS1|BS1\n!CS\n"),
                of("Z1A S13 \n Z1B\n", "!AS13\n!BS\n"),
                of("Z1A S13 | Z1B S1 \n Z1C\n", "!AS13\n!CS\n"));
    }

    @ParameterizedTest
    @MethodSource
    public void shouldProduceMultipleResponsesWithComments(final String text, String finalOutput) {
        parserActionTester.parseSnippet(text, finalOutput);
        assertThat(outStream.isOpen()).isFalse();
    }

    private static Stream<Arguments> shouldProduceMultipleResponsesWithComments() {
        return Stream.of(
                of("Z1 A \n#hi\n", "!AS\n!\n"),
                of("Z1 A \n#hi\n#abc\nZ1B\n", "!AS\n!\n!\n!BS\n"));
    }

    @ParameterizedTest
    @MethodSource
    public void shouldProduceMultipleResponsesWithAddressing(final String text, String finalOutput) {
        parserActionTester.parseSnippet(text, finalOutput);
        assertThat(outStream.isOpen()).isFalse();
    }

    private static Stream<Arguments> shouldProduceMultipleResponsesWithAddressing() {
        // Addressing actions are passed on downstream, not acknowledged - they should not appear in the response.
        return Stream.of(
                of("Z1 A \n@1 Z1B\n", "!AS\n"),
                of("Z1 A \n@1.2 Z1B\n@2.2 Z1C\n Z1D\n", "!AS\n!DS\n"));
    }
}
